src/Chapter3.hs

```diff
@@ -479,6 +551,52 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data Castle = NoCastle | CityCastle String deriving (Show)
+
+data Walls = NoWalls | CityWalls deriving (Show)
+
+data Hall = Church | Library deriving (Show)
+
+data House = BuildHouse 
+    {
+      housePoeple :: Int
+    } deriving (Show)
+
+data City = BuildCity
+    { cityCastle :: Castle
+    , cityHall :: Hall
+    , cityWalls :: Walls
+    , cityHouses :: [House]
+    } deriving (Show)
+
+type CastleName = String
+
+buildCastle :: City -> CastleName -> City
+buildCastle city name = city { cityCastle = (CityCastle name) }
+
+buildHouse :: City -> Int -> City
+buildHouse city numberOfPeople = city {cityHouses = freshHouse : (cityHouses city)}
+  where
+    freshHouse = BuildHouse numberOfPeople
+
+countPeople :: [House] -> Int
+countPeople houseArr = go 0 houseArr
+  where
+    go acc [] = acc
+    go acc (x:xs) = go (acc + (housePoeple x)) xs
```

Comment:
That's nice usage of a recursive function! Well done on using the "Recursive go" pattern. This pattern usually shines when you can't implement your function as a combination of standard functions. Here, you can use standard functions `map` and `sum` to implement a shorter version :slightly_smiling_face: